### ISSUE TITLE
 Middleware order

### ISSUE THREAD

**Issue by Frankofoedu** [IzyPro/WatchDog#35]
I believe the readme should warn new users to add the middleware in a particular order since it internally registers app.UseRouting(),
 app.UseEndpoints() and appUseMvc(). I had something similar to this 

              
```
            app.UseRouting();
            app.UseIdentityServer();
            app.UseAuthentication();
            app.UseAuthorization();

            app.UseHangfireDashboard("/ourhangfire", new DashboardOptions
            {
                Authorization = new[] {
                    new BasicAuthAuthorizationFilter(
                        new BasicAuthAuthorizationFilterOptions{
                            RequireSsl = false,
                            SslRedirect = false,
                            LoginCaseSensitive = true,
                            Users = new [] {
                                new BasicAuthAuthorizationUser {
                                    Login = "xxx@xx.com",
                                    PasswordClear =  "xxxx"
                                }
                            }
                        })
                }
            });


            app.UseWatchDog(opt =>
            {
                opt.WatchPageUsername = "xx@xx.xx";
                opt.WatchPagePassword = "xxx?";
            });

            app.UseEndpoints(endpoints =>
            {
                endpoints.MapControllers();
                endpoints.MapHangfireDashboard();
            });
```

which generated this error

> System.InvalidOperationException: Endpoint Web.Server.Controllers.Profile.ProfileController.UploadUsersPicture contains authorization metadata, but a middleware was not found that supports authorization.
> Configure your application startup by adding app.UseAuthorization() inside the call to Configure(..) in the application startup code. The call to app.UseAuthorization() must appear between app.UseRouting() and app.UseEndpoints(...).

Fixed by switching app.UseWatchDog() to the last line.

**Comment by grumpykiwi** [comment]
I have a similar possibly related problem.

If I include the setup calls for WatchDog in program.cs, nothing is displayed in my default view

```
using Microsoft.EntityFrameworkCore;
using Models;
using WatchDog;

var builder = WebApplication.CreateBuilder(args);

// Add services to the container.
builder.Services.AddControllersWithViews();

builder.Services.AddCors(options =>
{
	options.AddDefaultPolicy(
		policy =>
		{
			policy.AllowAnyOrigin().AllowAnyHeader().AllowAnyMethod();
		});
});

if (builder.Environment.EnvironmentName.ToLower() == "development" ||
	builder.Environment.EnvironmentName.ToLower() == "staging")
{
	builder.Services.AddWatchDogServices(opt =>
	{
		opt.IsAutoClear = true;
		opt.ClearTimeSchedule = WatchDog.src.Enums.WatchDogAutoClearScheduleEnum.Weekly;
	});
}

builder.Services.AddHttpClient();
builder.Services.AddDbContextFactory<DBContext>(options =>
	options.UseSqlServer(builder.Configuration.GetConnectionString("DBContext")));

var app = builder.Build();

// Configure the HTTP request pipeline.
if (!app.Environment.IsDevelopment())
{
	app.UseExceptionHandler("/Home/Error");
	app.UseHsts();
}
app.UseCors();

app.UseHttpsRedirection();

app.UseStaticFiles();

app.UseRouting();

app.UseAuthorization();

if (app.Environment.EnvironmentName.ToLower() == "development" || 
	app.Environment.EnvironmentName.ToLower() == "staging")
{
	app.UseWatchDogExceptionLogger();
	app.UseWatchDog(opt =>
	{
		opt.WatchPageUsername = "";
		opt.WatchPagePassword = "";
	});
}

app.MapControllerRoute(
	name: "default",
	pattern: "{controller=Home}/{action=Index}/{id?}");
app.Run();
```
View

```
<div class="text-center">
    <h1 class="display-4">Welcome</h1>
    <p>This is the landing page for PartnerCalc<sup>&trade;</sup>.</p>
    <p>This proves that routing is working and the controller engine is rendering views.</p>
</div>
```

Controller

```using Microsoft.AspNetCore.Mvc;
using Models;
using System.Diagnostics;

namespace Controllers;
public class HomeController : Controller
{
	public IActionResult Index()
	{
		return View();
	}
}
```
Any ideas. I am guessing it might have something to do with ordering. If I comment out the calls to Watchdog, the basic HTML will render correctly.

Thanks


Mark

**Comment by IzyPro** [comment]
Thanks @Frankofoedu and @grumpykiwi 
As correctly mentioned, ```app.UseWatchDog()``` should come after app.UseRouting(), app.UseEndpoints() and appUseMvc(). 
The readme has been updated to reflect that.

**Comment by grumpykiwi** [comment]
After a bit more fiddling I can get the home controller to return a string, but not any HTML. WatchDog seems to intercept it when requested by the browser. However I did notice that requesting the same URL from Postman https://localhost:7066/ actually returns the correct html when I return the HTML instead of a string

```
	public ContentResult Index()
	{
		var html = System.IO.File.ReadAllText(@"./views/home/index.cshtml");
		return base.Content(html, "text/html");
	}
```

See attached fil
![Watchdog](https://user-images.githubusercontent.com/43825887/168480122-7270119d-1de8-46c1-8c84-94c254662851.png)
e

Which is kind of weird. Hopefully that might help y'all in some way. Thanks

**Comment by IzyPro** [comment]
Hi @Frankofoedu @grumpykiwi 
Thanks for pointing these out
It has been fixed in the most recent version (v1.2.0) and you should be good to go after updating

**Comment by grumpykiwi** [comment]
Thanks very much. Updated and working perfectly. A great addition to my API toolkit. Cheers


Mark Nash
Senior Software Engineer        [cid:PC_42x23_fbb4e57e-65d2-4c41-b09e-ffa9f4a3c468.png]

PartnerComm, Inc.
Winner of 85 IABC Gold Quills
2021 Large Agency of the Year
2304 W Interstate 20, Suite 250
Arlington, TX 76017
        [cid:pcomm25+_471933a4-f842-490b-84d4-910cf0397a29.png]

Office: (817) 465-9277
***@***.***
www.partnercomm.net<http://www.partnercomm.net>





This e-mail transmission may contain confidential information or may be otherwise legally privileged. If you are not the intended recipient you are hereby notified that any use, disclosure, copying, forwarding, or distribution of any of the information contained in or attached to this message is STRICTLY PROHIBITED.

________________________________
